### PR TITLE
[Feat] Create settings page

### PR DESCRIPTION
**Pull request by kwajiehao**
### Note
This PR is to be evaluated in conjunction with PR#[48](https://github.com/isomerpages/isomercms-backend/pull/48) in the backend.

### Overview
This PR introduces a new `Settings` page which is accessible from the `Sidebar`. The `Settings` page will allow you to modify the following fields:
- Title
- Favicon
- Primary color
- Secondary color
- Media resource colors
- Resource room name
- Social media page links

The color fields (primary, secondary, and media resource colors) are equipped with a color picker which allows users to select colors directly. For now, we are using `SketchPicker` from the `react-color` package as our color picker. More information can be found [here](https://casesandberg.github.io/react-color/).

Field validation has been done for the social media fields.

### To-do
- Introduce error validation for remaining content fields
- Introduce WCAG2.0 validation for color fields
- Upload image functionality for image fields
- Further styling to refine appearance

### CODE REVIEW

**Comment by prestonlimlianjie** [comment]
There's some issue with the social media validation fields - I typed in some text in the social media fields and then deleted the text, but the error still remains.
Screenshot 2019-12-11 at 6 19 00 PM

**Comment by prestonlimlianjie** [comment]
The color picker also doesn't save the new color at times (even when I double click on a new colored pixel)

**Comment by prestonlimlianjie** [comment]
A feature to be added in the future: allowing users to add/delete media colors. Sites like MLAW have >10 colors in their media colors.

**Comment by kwajiehao** [comment]
@prestonlimlianjie Cleaned up some of the React errors and prevent saving if errors exist

File: src/layouts/Settings.jsx [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+    const flattenedErrors = Object.values(errors).reduce(this.retrieveErrorsCallback, []);
+    return _.some(flattenedErrors);
+  }
+
+  retrieveErrorsCallback = (acc, curr) => {
+    if (typeof curr === 'object') {
+      return acc.concat(this.retrieveErrors(curr));
+    }
+    return acc.concat(curr);
+  }
```
Comment (taufiqmmhd) [inline src/layouts/Settings.jsx]: Learnt that this is to check for any error values within a nested object. Don't think it's worth to do hard to read recursive functions just to check for error values within `media-colors`.

File: src/utils/validators.js [hunk 2]
```diff
@@ -2,6 +2,8 @@
 // Common regexes and constants
 // ==============
 const PERMALINK_REGEX = '^(/([a-z]+([-][a-z]+)*/)+)$';
+const SOCIAL_MEDIA_REGEX_PART_1 = '^(https://)?(www.)?(';
+const SOCIAL_MEDIA_REGEX_PART_2 = '.com/)[a-zA-Z0-9_-]+$';
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply src/utils/validators.js]: Solved in [ed10666](https://github.com/isomerpages/isomercms-frontend/pull/99/commits/ed1066639bf27e143e66146613f972ecc7020bf1)

File: src/layouts/Settings.jsx [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+                {/* Favicon field */}
+                <FormFieldImage
+                  title="Favicon"
+                  id="favicon"
+                  value={favicon}
+                  errorMessage={errors.favicon}
+                  isRequired
+                  onFieldChange={this.changeHandler}
+                  onColorClick={this.colorPickerToggle}
+                />
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply src/layouts/Settings.jsx]: image selection to be done in a subsequent PR

File: src/layouts/Settings.jsx [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+  realTimeColor = (color) => {
+    const { colorPicker: { elementId }, colors } = this.state;
+    // there is no hex property if the color submitted is the
+    // same as the original color
+    const hex = color.hex ? color.hex : `#${color}`;
+
+    // set state of color fields
+    if (elementId === 'primary-color' || elementId === 'secondary-color') {
+      this.setState((currState) => ({
+        ...currState,
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply src/layouts/Settings.jsx]: we need to do this unless we use `immutability-helper', which we will implement at a later date

File: src/layouts/Settings.jsx [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+          [elementId]: hex,
+        },
+      }));
+    } else {
+      // set state of resource colors
+      const index = elementId.split('@')[elementId.split('@').length - 1];
+      const newMediaColors = _.cloneDeep(colors['media-colors']);
+      newMediaColors[index].color = hex;
+      this.setState((currState) => ({
+        ...currState,
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply src/layouts/Settings.jsx]: see above

File: src/layouts/Settings.jsx [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+        colors: {
+          ...currState.colors,
+          'media-colors': newMediaColors,
+        },
+      }));
+    } else {
+      this.setState((currState) => ({
+        // errors: {
+        //   [id]: errorMessage,
+        // },
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply src/layouts/Settings.jsx]: resolved with [85481e](https://github.com/isomerpages/isomercms-frontend/pull/99/commits/85481e521928b3baaad77e2136d054d8c97092a3)

File: src/layouts/Settings.jsx [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+
+  // onColorSelect sets value of appropriate color field
+  onColorSelect = (event, color) => {
+    // prevent event from reloading
+    // prevent parent form from being submitted
+    event.preventDefault();
+    event.stopPropagation();
+
+    // reflect color changes
+    this.realTimeColor(color);
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply src/layouts/Settings.jsx]: resolved with [85481e](https://github.com/isomerpages/isomercms-frontend/pull/99/commits/85481e521928b3baaad77e2136d054d8c97092a3)

File: src/layouts/Settings.jsx [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+    }
+  }
+
+  disableColorPicker = (resetOldColors) => {
+    const { colorPicker: { colorPickerToggle, oldColors } } = this.state;
+    // if ColorPicker is active, disable it
+    if (colorPickerToggle) {
+      if (resetOldColors) {
+        this.setState((currState) => ({
+          ...currState,
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply src/layouts/Settings.jsx]: resolved with [85481e](https://github.com/isomerpages/isomercms-frontend/pull/99/commits/85481e521928b3baaad77e2136d054d8c97092a3)

File: src/layouts/Settings.jsx (hunk 1 above)
Comment (kwajiehao) [reply src/layouts/Settings.jsx]: Fixed with [50fd1f6c](https://github.com/isomerpages/isomercms-frontend/pull/99/commits/50fd1f6cce5e68395c71d8c2bc92340bee022e68)

**Pull request merged**